src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs

```diff
@@ -0,0 +1,79 @@
+﻿using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Options;
+
+namespace Workleap.Extensions.Http.Authentication.ClientCredentialsGrant;
+
+internal sealed class CacheTokenOnStartupBackgroundService : BackgroundService
+{
+    private readonly IHostApplicationLifetime _applicationLifetime;
+    private readonly IClientCredentialsTokenManagementService _tokenManagementService;
+    private readonly IOptions<CacheTokenOnStartupBackgroundServiceOptions> _backgroundServiceOptions;
+    private readonly SemaphoreSlim _allTokensCachedSignal;
+    private readonly List<string> _clientNames;
+
+    private CancellationTokenRegistration? _applicationStartedRegistration;
+    private int _successfulCachedTokenCount;
+
+    public CacheTokenOnStartupBackgroundService(
+        IHostApplicationLifetime applicationLifetime,
+        IClientCredentialsTokenManagementService tokenManagementService,
+        IOptions<CacheTokenOnStartupBackgroundServiceOptions> backgroundServiceOptions)
+    {
+        this._applicationLifetime = applicationLifetime;
+        this._tokenManagementService = tokenManagementService;
+        this._backgroundServiceOptions = backgroundServiceOptions;
+        this._allTokensCachedSignal = new SemaphoreSlim(initialCount: 0, maxCount: 1);
+        this._clientNames = new List<string>();
+    }
+
+    protected override Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        this._clientNames.AddRange(this._backgroundServiceOptions.Value.ClientCredentialPoweredClientNames);
+
+        this._applicationStartedRegistration = this._applicationLifetime.ApplicationStarted.Register(() =>
```

Comment:
1. We know for sure ([host source code](https://github.com/dotnet/runtime/blob/v8.0.5/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L159)) that when `IHostApplicationLifetime.ApplicationStarted` is triggered, the host is fully started (all IHostedService have been executed, web server is up and running)
2. If we were to execute this code directly in the `BackgroundService.ExecuteAsync` method, then the web server wouldn't be started yet. While this is not an issue with actual production code (Kestrel is not up yet, no problem), it does crash in integration tests because the testing server is not available yet but is requested by the integration tests which attempts to react to our "token are cached" and tries to access the web app (which is still starting at this point)

TLDR: it was required for integration test, but overall it's better to listen to this signal because we know for a fact that at this point the server is up and running